rollup/pipeline/pipeline.go

```diff
@@ -0,0 +1,407 @@
+			txs.Pop()
+		}
+	}
+
+	return nil
+}
+
+func (p *Pipeline) TryPushTxn(tx *types.Transaction) (*Result, error) {
+	select {
+	case p.TxnQueue <- tx:
+	case res := <-p.ResultCh:
+		return res, nil
+	}
+
+	select {
+	case err, valid := <-p.applyStageRespCh:
+		if !valid {
+			return nil, ErrApplyStageDone
+		}
+		return nil, err
+	case res := <-p.ResultCh:
+		return res, nil
+	}
+}
+
+func (p *Pipeline) Kill() {
+	if p.TxnQueue != nil {
+		close(p.TxnQueue)
+	}
+
+	select {
+	case <-p.applyStageRespCh:
+		<-p.ResultCh
+	case <-p.ResultCh:
+		<-p.applyStageRespCh
+	}
+}
+
+type PendingBlockIncrement struct {
+	LastTrace      *types.BlockTrace
+	NextL1MsgIndex uint64
+
+	// accumulated state
+	Header        *types.Header
+	State         *state.StateDB
+	Txs           types.Transactions
+	Receipts      types.Receipts
+	CoalescedLogs []*types.Log
+}
+
+func (p *Pipeline) traceAndApplyStage(txsIn <-chan *types.Transaction) (chan error, chan *PendingBlockIncrement, error) {
+	p.state.StartPrefetcher("miner")
+	incrementCh := make(chan *PendingBlockIncrement)
+	resCh := make(chan error)
+	go func() {
+		defer func() {
+			close(incrementCh)
+			close(resCh)
+			p.state.StopPrefetcher()
+		}()
+
+		var tx *types.Transaction
+		for {
+			applyIdleTimer.Time(func() {
+				tx = <-txsIn
+			})
+			if tx == nil {
+				return
+			}
+
+			applyStart := time.Now()
+
+			// If we don't have enough gas for any further transactions then we're done
+			if p.gasPool.Gas() < params.TxGas {
+				return
+			}
+
+			// If we have collected enough transactions then we're done
+			// Originally we only limit l2txs count, but now strictly limit total txs number.
+			if !p.chain.Config().Scroll.IsValidTxCount(p.txs.Len() + 1) {
+				return
+			}
+
+			if tx.IsL1MessageTx() && tx.AsL1MessageTx().QueueIndex != p.nextL1MsgIndex {
+				// Continue, we might still be able to include some L2 messages
+				resCh <- ErrUnexpectedL1MessageIndex
+				continue
+			}
+
+			if !tx.IsL1MessageTx() && !p.chain.Config().Scroll.IsValidBlockSize(p.blockSize+tx.Size()) {
+				// can't fit this txn in this block, silently ignore and continue looking for more txns
+				resCh <- nil
+				continue
+			}
+
+			// Start executing the transaction
+			p.state.SetTxContext(tx.Hash(), p.txs.Len())
+			receipt, trace, err := p.traceAndApply(tx)
+
+			if p.txs.Len() == 0 && tx.IsL1MessageTx() && err != nil {
+				// L1 message errored as the first txn, skip
+				p.nextL1MsgIndex = tx.AsL1MessageTx().QueueIndex + 1
+			}
+
+			if err == nil {
+				// Everything ok, collect the logs and shift in the next transaction from the same account
+				p.coalescedLogs = append(p.coalescedLogs, receipt.Logs...)
+				p.txs = append(p.txs, tx)
+				p.receipts = append(p.receipts, receipt)
+
+				if !tx.IsL1MessageTx() {
+					// only consider block size limit for L2 transactions
+					p.blockSize += tx.Size()
+				} else {
+					p.nextL1MsgIndex = tx.AsL1MessageTx().QueueIndex + 1
+				}
+
+				stallStart := time.Now()
```

Comment:
We can consider moving the commit phase to here as well. Instead of stalling to wait for CCC stage, we can calculate the state commitment-per-txn here, effectively making commit phase entirely free.